docs/zkapps/o1js/bitwise-operations.mdx

```diff
@@ -0,0 +1,456 @@
+---
+title: Bitwise Operations
+hide_title: true
+description: Bitwise operations in o1js are implemented as gadgets.
+keywords:
+  - gadgets
+  - bitwise operations
+  - o1js
+---
+
+:::info
+
+Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.
+
+:::
+
+# Bitwise Operations
+
+In o1js, bitwise operations are implemented as [gadgets](/zkapps/o1js/gadgets).
+
+- [and()](#and)
+- [not()](#not)
+- [xor()](#xor)
+- [addMod32()](#addmod32)
+- [divMod32()](#divmod32)
+- [leftShift32()](#leftshift32)
+- [leftShift64()](#leftshift64)
+- [rightShift32()](#rightshift32)
+- [rightShift64()](#rightshift64)
+- [rotate32()](#rotate32)
+- [rotate64()](#rotate64)
+- [rangeCheck32()](#rangecheck32)
+- [rangeCheck64()](#rangecheck64)
+- [multiRangeCheck()](#multirangecheck)
+- [compactMultiRangeCheck()](#compactmultirangecheck)
+
+## and()
+
+```ts
+and(a: Field, b: Field, length: number) => Field
+```
+
+The bitwise `and()` gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 
```

Comment:
Is `and()` a provable method? If yes, let's add

Reply:
is and() only for native Field elements? (or maybe all of these bitwise operations are for native field elements?)
I feel like we can clarify

Reply:
They are indeed for Field elements. This can be seen in [the function definition](https://github.com/o1-labs/o1js/blob/4efb9e505ba12170cbc10c372815e9532aafc77f/src/lib/gadgets/bitwise.ts#L153). It accepts Field types as inputs. And yes anything provable has to work on Field types so this applies to all of these operations. We only need to reimplement them because they have to be provable. They are well established and already exist in a non-zk context / not on Field types.

Reply:
* all of these are provable methods, so maybe good to highlight that more broadly in the introduction, instead of on each method. (re what is it proving - it's proving that you computed the bitwise AND of two numbers)
* all of these operate on native field elements (`Field`). I wouldn't necessarily highlight that on each method because `Field` is kind of the "default type" in provable code